src/dune_rules/preprocessing.ml

```diff
@@ -536,65 +536,174 @@ let driver_flags expander ~corrected_suffix ~driver_flags ~standard =
-                    ~scope pps
-                in
-                let+ ppx_flags =
-                  driver_flags expander ~corrected_suffix
-                    ~driver_flags:driver.info.lint_flags
-                    ~standard:(Action_builder.return [])
-                in
-                (exe, ppx_flags, flags))
+  @@
+  let alias = Alias.lint ~dir in
+  let add_alias build = Super_context.add_alias_action sctx alias build ~dir in
+  let lint =
+    Module_name.Per_item.map lint ~f:(function
+      | Preprocess.No_preprocessing -> fun ~source:_ ~ast:_ -> Memo.return ()
+      | Future_syntax loc ->
+        User_error.raise ~loc [ Pp.text "'compat' cannot be used as a linter" ]
+      | Action (loc, action) ->
+        fun ~source ~ast:_ ->
+          Module.iter source ~f:(fun _ (src : Module.File.t) ->
+              let src = Path.as_in_build_dir_exn (Module.File.path src) in
+              add_alias ~loc:(Some loc)
+                (action_for_pp ~sandbox ~loc ~expander ~action ~src))
+      | Pps { loc; pps; flags; staged } ->
+        if staged then
+          User_error.raise ~loc
+            [ Pp.text "Staged ppx rewriters cannot be used as linters." ];
+        let corrected_suffix = ".lint-corrected" in
+        let driver_and_flags =
+          Action_builder.memoize
+            ~cutoff:
+              (Tuple.T3.equal Path.Build.equal (List.equal String.equal)
+                 (List.equal String.equal))
+            "ppx driver and flags"
+            (let* () = Action_builder.return () in
+             let* exe, driver, flags =
+               ppx_driver_and_flags sctx ~expander ~loc ~lib_name ~flags ~scope
+                 pps
+             in
+             let+ ppx_flags =
+               driver_flags expander ~corrected_suffix
+                 ~driver_flags:driver.info.lint_flags
+                 ~standard:(Action_builder.return [])
+             in
+             (exe, ppx_flags, flags))
+        in
+        fun ~source ~ast ->
+          Module.iter ast ~f:(fun ml_kind src ->
+              add_alias ~loc:(Some loc)
+                (promote_correction ~suffix:corrected_suffix
+                   (Path.as_in_build_dir_exn
+                      (Option.value_exn (Module.file source ~ml_kind)))
+                   (let* exe, flags, args = driver_and_flags in
+                    let dir =
+                      Path.build (Super_context.context sctx).build_dir
+                    in
+                    Command.run' ~dir
+                      (Ok (Path.build exe))
+                      [ As args
+                      ; Command.Ml_kind.ppx_driver_flag ml_kind
+                      ; Dep (Module.File.path src)
+                      ; As flags
+                      ]))))
+  in
+  fun ~(source : Module.t) ~ast ->
+    Module_name.Per_item.get lint (Module.name source) ~source ~ast
+
+let pp_one_module sctx ~lib_name ~scope ~preprocessor_deps
```

Comment:
Is this only used by `make`? If so, wouldn't it be better for it to go there?

Reply:
Because this allows us to decrease like 3 levels of indentation. Moreover, it's still a private function.